### ISSUE TITLE
[Bug]: AST parsing fails on call to map key on an object

### ISSUE THREAD

**Issue by maxxxxxdlp** [jsx-eslint/eslint-plugin-react#3516]
### Is there an existing issue for this?

- [X] I have searched the existing issues and my issue is unique
- [X] My issue appears in the command-line and not only in the text editor

### Description Overview

AST parsing fails on call to `map` key on an object.

Happens for the following `.tsx` file:
```tsx
import { wbPlanText } from '../../localization/wbPlan';

wbPlanText.map();
```

Running `npx eslint lib/components/WbPlanView/Mapper.tsx` fails:

```yaml
Oops! Something went wrong! :(

ESLint: 8.31.0

TypeError: Cannot read properties of undefined (reading 'type')
Occurred while linting /Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/lib/components/WbPlanView/Mapper.tsx:3
Rule: "react/jsx-key"
    at Object.isFunctionLikeExpression (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint-plugin-react/lib/util/ast.js:223:15)
    at CallExpression[callee.type="MemberExpression"][callee.property.name="map"],       CallExpression[callee.type="OptionalMemberExpression"][callee.property.name="map"],       OptionalCallExpression[callee.type="MemberExpression"][callee.property.name="map"],       OptionalCallExpression[callee.type="OptionalMemberExpression"][callee.property.name="map"] (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint-plugin-react/lib/rules/jsx-key.js:230:22)
    at ruleErrorHandler (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/linter.js:1115:28)
    at /Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/node-event-generator.js:297:26)
    at NodeEventGenerator.applySelectors (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/node-event-generator.js:326:22)
    at NodeEventGenerator.enterNode (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/node-event-generator.js:340:14)
    at CodePathAnalyzer.enterNode (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/code-path-analysis/code-path-analyzer.js:795:23)
```

Changing the last line to `wbPlanText.maps();` fixes the bug.

### Expected Behavior

No error

### eslint-plugin-react version

7.31.11

### eslint version

8.31.0

### node version

v19.3.0

**Comment by maxxxxxdlp** [comment]
Another example:

```ts
  function AutoMapper(
    parameters: AutoMapperConstructorParameters
  ): AutoMapperResults {
    return new AutoMapperConstructor(parameters).map();
  },
```

```yaml
TypeError: Cannot read properties of undefined (reading 'type')
Occurred while linting /Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/lib/components/WbPlanView/__tests__/automapper.test.ts:16
Rule: "react/jsx-key"
    at Object.isFunctionLikeExpression (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint-plugin-react/lib/util/ast.js:223:15)
    at CallExpression[callee.type="MemberExpression"][callee.property.name="map"],       CallExpression[callee.type="OptionalMemberExpression"][callee.property.name="map"],       OptionalCallExpression[callee.type="MemberExpression"][callee.property.name="map"],       OptionalCallExpression[callee.type="OptionalMemberExpression"][callee.property.name="map"] (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint-plugin-react/lib/rules/jsx-key.js:230:22)
    at ruleErrorHandler (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/linter.js:1115:28)
    at /Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/node-event-generator.js:297:26)
    at NodeEventGenerator.applySelectors (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/node-event-generator.js:326:22)
    at NodeEventGenerator.enterNode (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/node-event-generator.js:340:14)
    at CodePathAnalyzer.enterNode (/Users/maxpatiiuk/site/python/specify7/specifyweb/frontend/js_src/node_modules/eslint/lib/linter/code-path-analysis/code-path-analyzer.js:795:23)
```

**Comment by ljharb** [comment]
Thanks - presumably it still happens when you pass an argument to `.map` (it'd be really really confusing for any API in JS to have a function named "map" that didn't take a function and do a mapping).

**Comment by maxxxxxdlp** [comment]
Good suggestion. The exception does not happen when I modify the above line to:

```ts
    return new AutoMapperConstructor(parameters).map('test');
```

**Comment by ljharb** [comment]
Hmm, I can't reproduce this. What version of eslint-plugin-react do you have?

**Comment by ljharb** [comment]
What parser are you using?

**Comment by maxxxxxdlp** [comment]
If it helps to reproduce this, here is my `.eslintrc.js`:

```
require('@rushstack/eslint-patch/modern-module-resolution');

const OFF = 'off';
const WARN = 'warn';
const ERROR = 'error';

module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: './tsconfig.eslint.json',
  },
  env: {
    browser: true,
    node: true,
  },
  extends: ['@maxxxxxdlp/eslint-config-react'],
  rules: {
    '@typescript-eslint/no-empty-interface': OFF,
  },
};
```

The parser is `@typescript-eslint/parser`


Relevant dependencies:
```
"@maxxxxxdlp/eslint-config-react": "^1.0.2",
"@rushstack/eslint-patch": "^1.2.0",
"eslint": "^8.31.0",

```

**Comment by ljharb** [comment]
I'm very skeptical that eslint-patch thing has no side effects - are you sure that's not causing issues? What's `npm explain eslint-plugin-react` output?

**Comment by ljharb** [comment]
Thanks, the repro repo really helps, i'll take a look.

**Comment by ljharb** [comment]
Definitely the eslint-patch hack is causing problems; normally i can `npm link` the plugin in and reproduce it with my checked-out project, i get this error:
```
Oops! Something went wrong! :(

ESLint: 8.20.0

ESLint couldn't determine the plugin "react" uniquely.

- $reactPlugin/index.js (loaded in ".eslintrc.js » @maxxxxxdlp/eslint-config-react")
- null (loaded in ".eslintrc.js » @maxxxxxdlp/eslint-config-react » plugin:react/recommended")

Please remove the "plugins" setting from either config or remove either plugin installation.

If you still can't figure out the problem, please stop by https://eslint.org/chat/help to chat with the team.
```

I really suggest just using peer deps properly instead of trying to hack around it :-/

I'll keep trying to figure out how to repro it with a linked project tho.

**Comment by ljharb** [comment]
Actually, when I run the node repl in that dir and do `require('eslint-plugin-react/package.json').version`, i get `v7.31.1`, not `v7.31.11`, so I think that's the issue. Can you update to the latest version of the plugin?

**Comment by maxxxxxdlp** [comment]
Can confirm that the issue is fixed in `v7.31.11`

Sorry for the confusion. I should have been more careful when copying the version number

**Comment by ljharb** [comment]
Glad to hear it :-)
